common/src/main/java/io/pravega/common/util/btree/ByteArrayComparator.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util.btree;
+
+import io.pravega.common.util.ByteArraySegment;
+import java.io.Serializable;
+import java.util.Comparator;
+
+/**
+ * Compares two byte arrays of the same length.
+ */
+class ByteArrayComparator implements Comparator<byte[]>, Serializable {
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public int compare(byte[] b1, byte[] b2) {
+        assert b1.length == b2.length;
+        return compare(b1, 0, b2, 0, b1.length);
+    }
+
+    /**
+     * Compares two non-null ByteArraySegments of the same length.
+     *
+     * @param b1 First instance.
+     * @param b2 Second instance.
+     * @return -1 if b1 should be before b2, 0 if b1 equals b2 and 1 if b1 should be after b2.
+     */
+    int compare(ByteArraySegment b1, ByteArraySegment b2) {
+        assert b1.getLength() == b2.getLength();
+        return compare(b1.array(), b1.arrayOffset(), b2.array(), b2.arrayOffset(), b1.getLength());
+    }
+
+    /**
+     * Compares two byte arrays from the given offsets.
+     *
+     * @param b1      First array.
+     * @param offset1 The first offset to begin comparing in the first array.
+     * @param b2      Second array.
+     * @param offset2 The first offset to begin comparing in the second array.
+     * @param length  The number of bytes to compare.
+     * @return -1 if b1 should be before b2, 0 if b1 equals b2 and 1 if b1 should be after b2.
+     */
+    int compare(byte[] b1, int offset1, byte[] b2, int offset2, int length) {
+        int r;
+        for (int i = 0; i < length; i++) {
```

Comment:
You can make this slightly faster by having offset1 and offset2 directly incremented by the loop, so as to eliminate i, and adding a precondition check at the top to make sure that length is not too long. (This will allow the JIT to remove the outOfBounds check on the array access.)

Reply:
Hmm Interesting. I have tried your approach (and one more), and when changing the unit test for this to do 100M comparisons (instead of the regular 10000), i noticed no runtime difference between the differnt approaches. 100M comparisons took 1200ms for any sort of implementation.